### ISSUE TITLE
Modals (or others) with component as content does not trigger ngOnChanges

### ISSUE THREAD

**Issue by MickL** [ng-bootstrap/ng-bootstrap#2146]
This is really nice:

```
const modalRef = this.modalService.open(NgbdModalContent);
modalRef.componentInstance.name = 'World';
```

Anyway this does not trigger `ngOnChanges()` when we set the input "name" but this lifecycle-hook are a crucial feature of Angular components.

**Comment by pkozlowski-opensource** [comment]
Unfortunately this is how Angular works with any dynamically created component. We can't do much about it here, in this project, it is not specific to ng-bootstrap.

**Comment by ip-amity-tarun** [comment]
private cd_interval_fn: any;
constructor(private cdref: ChangeDetectorRef ) {
    this.cdref.detach();
    this.cd_interval_fn = setInterval(() => {
      this.cdref.detectChanges();
    }, 500);
  }
ngOnDestroy() {
    clearInterval(this.cd_interval_fn);
    this.cdref.detach();
  }

**Comment by OdeToSpot** [comment]
I was passing values to an ngbModalRef, and it needed the values to make a request to an API, and since ngOnChanages() wasn't firing, I came up with this work-around.  Essentially the modal checks if the value is there and does something.  If not, it waits a second and either dismisses with an error or does something.

Parent component
```
const modalRef = this.modalService.open(NgbdModalContent);
modalRef.componentInstance.name = 'World';
```
Child modal component
```
ngOnInit() {
  if (this.name) {
    this.doSomethingWIthName();
  } else {
   setTimeout(() => { 
     if (this.name) { this.doSomethingWithName() }
     else { activeModal.dismiss("I never got a name...")}, 1000);
}
```

**Comment by IAfanasov** [comment]
one option
``` TS
protected _name: string;
get name():string { return this._name; }
set name(value:string) {  this._name = value; this.cdref.detectChanges(); }
constructor(protected cdref: ChangeDetectorRef ) { ... }
```
other
``` TS
// component inside modal
public name: string;
constructor(public cdref: ChangeDetectorRef ) { ... }

// component opening modal
const modalRef = this.modalService.open(NgbdModalContent);
modalRef.componentInstance.name = 'World';
modalRef.componentInstance.cdref.detectChanges();
```

Instead of `detectChanges` would be enough to use `markForCheck` if you use the OnPush change detection strategy

**Comment by dstj** [comment]
@IAfanasov , your proposed workaround didn't work for me. `ngOnChanges` still wouldn't get triggered. :/

The only workaround I got working was to wrap the component with `ngOnChanges` in another component that only forwards the binding.

Like this:

```typescript
const modalRef = this.modalService.open(ModalContentComponent);
modalRef.componentInstance.name = 'World';

...

@Component({
	selector: 'app-modal-content-modal',
	template: `<app-inner-modal-content [name]="name"></app-inner-modal-content>`,
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ModalContentComponent {
	@Input() name: string;
}

@Component({
	selector: 'app-inner-modal-content-modal',
	template: `...`,
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class InnerModalContentComponent {
	@Input() name: string;
	...
}
```

Using Angular v7.2.15 (if ever that changes anything)

**Comment by JAspeling** [comment]
It's not pretty, but @dstj 's answer worked for me, where the content of the modal is just wrapped inside the actual component used by `ng-bootstrap` to launch the modal.

**Comment by Stevetanus** [comment]
According to this [article](https://netbasal.com/a-new-way-to-set-inputs-on-angular-componentrefs-6214f95db63d), we can now use `setInput` build in method to solve the problem of `ngOnChanges` does not trigger.
